In [1]:
import pandas as pd
import numpy as np
import math

In [2]:
tennis=pd.read_csv('play_tennis.csv')
iris=pd.read_csv('Iris.csv')

## Part 1

In [3]:
tennis

day   outlook  temp humidity    wind play
0    D1     Sunny   Hot     High    Weak   No
1    D2     Sunny   Hot     High  Strong   No
2    D3  Overcast   Hot     High    Weak  Yes
3    D4      Rain  Mild     High    Weak  Yes
4    D5      Rain  Cool   Normal    Weak  Yes
5    D6      Rain  Cool   Normal  Strong   No
6    D7  Overcast  Cool   Normal  Strong  Yes
7    D8     Sunny  Mild     High    Weak   No
8    D9     Sunny  Cool   Normal    Weak  Yes
9   D10      Rain  Mild   Normal    Weak  Yes
10  D11     Sunny  Mild   Normal  Strong  Yes
11  D12  Overcast  Mild     High  Strong  Yes
12  D13  Overcast   Hot   Normal    Weak  Yes
13  D14      Rain  Mild     High  Strong   No

### Entropy of the parent 

In [4]:
round(-9/14*math.log(9/14,2)-5/14*math.log(5/14,2),2)

0.94

In [5]:
df1=tennis[tennis['outlook']=='Sunny']
df2=tennis[tennis['outlook']=='Overcast']
df3=tennis[tennis['outlook']=='Rain']

In [6]:
df1.reset_index()

index  day outlook  temp humidity    wind play
0      0   D1   Sunny   Hot     High    Weak   No
1      1   D2   Sunny   Hot     High  Strong   No
2      7   D8   Sunny  Mild     High    Weak   No
3      8   D9   Sunny  Cool   Normal    Weak  Yes
4     10  D11   Sunny  Mild   Normal  Strong  Yes

In [7]:
#Entropy of outlook being sunny

round(-(2/5*math.log2(2/5))-(3/5*math.log2(3/5)),2)

0.97

In [8]:
df2.reset_index()

index  day   outlook  temp humidity    wind play
0      2   D3  Overcast   Hot     High    Weak  Yes
1      6   D7  Overcast  Cool   Normal  Strong  Yes
2     11  D12  Overcast  Mild     High  Strong  Yes
3     12  D13  Overcast   Hot   Normal    Weak  Yes

In [9]:
#Entropy of outlook being overcast

0 #as all are yes

0

In [10]:
df3.reset_index()

index  day outlook  temp humidity    wind play
0      3   D4    Rain  Mild     High    Weak  Yes
1      4   D5    Rain  Cool   Normal    Weak  Yes
2      5   D6    Rain  Cool   Normal  Strong   No
3      9  D10    Rain  Mild   Normal    Weak  Yes
4     13  D14    Rain  Mild     High  Strong   No

In [11]:
#Entropy of outlook being rain

round(-3/5*math.log(3/5,2)-2/5*math.log(2/5,2),2)

0.97

### Weighted Entropy of outlook col.

In [12]:
round(5/14*0.97+4/14*0+5/14*0.97,2)

0.69

### Information Gain when we split data on the basis of outlook col .

In [13]:
ig1=0.94-0.69
ig1

0.25

In [14]:
df1=tennis[tennis['temp']=='Hot']
df2=tennis[tennis['temp']=='Mild']
df3=tennis[tennis['temp']=='Cool']

In [15]:
df1.reset_index()

index  day   outlook temp humidity    wind play
0      0   D1     Sunny  Hot     High    Weak   No
1      1   D2     Sunny  Hot     High  Strong   No
2      2   D3  Overcast  Hot     High    Weak  Yes
3     12  D13  Overcast  Hot   Normal    Weak  Yes

In [16]:
#Entropy of temp being hot

round(-2/4*math.log(2/4,2)-2/4*math.log(2/4,2),2)

1.0

In [17]:
df2.reset_index()

index  day   outlook  temp humidity    wind play
0      3   D4      Rain  Mild     High    Weak  Yes
1      7   D8     Sunny  Mild     High    Weak   No
2      9  D10      Rain  Mild   Normal    Weak  Yes
3     10  D11     Sunny  Mild   Normal  Strong  Yes
4     11  D12  Overcast  Mild     High  Strong  Yes
5     13  D14      Rain  Mild     High  Strong   No

In [18]:
#Entropy of temp being mild

round(-4/6*math.log(4/6,2)-2/6*math.log(2/6,2),2)

0.92

In [19]:
df3.reset_index()

index day   outlook  temp humidity    wind play
0      4  D5      Rain  Cool   Normal    Weak  Yes
1      5  D6      Rain  Cool   Normal  Strong   No
2      6  D7  Overcast  Cool   Normal  Strong  Yes
3      8  D9     Sunny  Cool   Normal    Weak  Yes

In [20]:
#Entropy of temp being cool

round(-3/4*math.log(3/4,2)-1/4*math.log(1/4,2),2)

0.81

### Weighted Entropy of temp col.

In [21]:
round(4/14*1.0+6/14*0.92+4/14*0.81,2)

0.91

### Information Gain when we split data on the basis of temp col

In [22]:
ig2=round(0.94-0.91,2)
ig2

0.03

In [23]:
df1=tennis[tennis['humidity']=='High']
df2=tennis[tennis['humidity']=='Normal']

In [24]:
df1.reset_index()

index  day   outlook  temp humidity    wind play
0      0   D1     Sunny   Hot     High    Weak   No
1      1   D2     Sunny   Hot     High  Strong   No
2      2   D3  Overcast   Hot     High    Weak  Yes
3      3   D4      Rain  Mild     High    Weak  Yes
4      7   D8     Sunny  Mild     High    Weak   No
5     11  D12  Overcast  Mild     High  Strong  Yes
6     13  D14      Rain  Mild     High  Strong   No

In [25]:
#Entropy of humidity being high

round(-3/7*math.log(3/7,2)-4/7*math.log(4/7,2),2)

0.99

In [26]:
df2.reset_index()

index  day   outlook  temp humidity    wind play
0      4   D5      Rain  Cool   Normal    Weak  Yes
1      5   D6      Rain  Cool   Normal  Strong   No
2      6   D7  Overcast  Cool   Normal  Strong  Yes
3      8   D9     Sunny  Cool   Normal    Weak  Yes
4      9  D10      Rain  Mild   Normal    Weak  Yes
5     10  D11     Sunny  Mild   Normal  Strong  Yes
6     12  D13  Overcast   Hot   Normal    Weak  Yes

In [27]:
#Entropy of humodity being normal

round(-6/7*math.log(6/7,2)-1/7*math.log(1/7,2),2)

0.59

### Weighted Entropy of humidity col

In [28]:
round(7/14*0.99+7/14*0.59,2)

0.79

### Information Gain when we split data on the basis of humidity col

In [29]:
ig3=round(0.94-0.79,2)
ig3

0.15

In [30]:
df1=tennis[tennis['wind']=='Weak']
df2=tennis[tennis['wind']=='Strong']

In [31]:
df1.reset_index()

index  day   outlook  temp humidity  wind play
0      0   D1     Sunny   Hot     High  Weak   No
1      2   D3  Overcast   Hot     High  Weak  Yes
2      3   D4      Rain  Mild     High  Weak  Yes
3      4   D5      Rain  Cool   Normal  Weak  Yes
4      7   D8     Sunny  Mild     High  Weak   No
5      8   D9     Sunny  Cool   Normal  Weak  Yes
6      9  D10      Rain  Mild   Normal  Weak  Yes
7     12  D13  Overcast   Hot   Normal  Weak  Yes

In [32]:
#Entropy of wind being weak

round(-6/8*math.log(6/8,2)-2/8*math.log(2/8,2),2)

0.81

In [33]:
df2.reset_index()

index  day   outlook  temp humidity    wind play
0      1   D2     Sunny   Hot     High  Strong   No
1      5   D6      Rain  Cool   Normal  Strong   No
2      6   D7  Overcast  Cool   Normal  Strong  Yes
3     10  D11     Sunny  Mild   Normal  Strong  Yes
4     11  D12  Overcast  Mild     High  Strong  Yes
5     13  D14      Rain  Mild     High  Strong   No

In [34]:
#Entropy of wind being strong

round(-3/6*math.log(3/6,2)-3/6*math.log(3/6,2),2)

1.0

### Weighted entropy of wind col

In [35]:
round(8/14*0.81+6/14*1.0,2)

0.89

### Information Gain when we split data on the basis of wind col.

In [36]:
ig4=round(0.94-0.89,2)
ig4

0.05

In [37]:
ig1 #outlook

0.25

In [38]:
ig2 #temp

0.03

In [39]:
ig3 #humidity

0.15

In [40]:
ig4 #wind

0.05

In [41]:
#conclusion

### As the Information gain of outlook column is the greatest therefore the first spliting will be on the outlook column

# Part 2

In [42]:
iris

Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm  \
0      1            5.1           3.5            1.4           0.2   
1      2            4.9           3.0            1.4           0.2   
2      3            4.7           3.2            1.3           0.2   
3      4            4.6           3.1            1.5           0.2   
4      5            5.0           3.6            1.4           0.2   
..   ...            ...           ...            ...           ...   
145  146            6.7           3.0            5.2           2.3   
146  147            6.3           2.5            5.0           1.9   
147  148            6.5           3.0            5.2           2.0   
148  149            6.2           3.4            5.4           2.3   
149  150            5.9           3.0            5.1           1.8   

            Species  
0       Iris-setosa  
1       Iris-setosa  
2       Iris-setosa  
3       Iris-setosa  
4       Iris-setosa  
..              ...  
145  Iris-virginica  
146  Iris-virginica  
147  Iris-virginica  
148  Iris-virginica  
149  Iris-virginica  

[150 rows x 6 columns]

In [43]:
iris.rename(columns={'PetalLengthCm':'PL','SepalLengthCm':'SL','SepalWidthCm':'SW','PetalWidthCm':'PW'},inplace=True)

In [44]:
df=iris[['PL','Species']]

In [45]:
df['Species'].value_counts()

Iris-versicolor    50
Iris-setosa        50
Iris-virginica     50
Name: Species, dtype: int64

In [46]:
df

PL         Species
0    1.4     Iris-setosa
1    1.4     Iris-setosa
2    1.3     Iris-setosa
3    1.5     Iris-setosa
4    1.4     Iris-setosa
..   ...             ...
145  5.2  Iris-virginica
146  5.0  Iris-virginica
147  5.2  Iris-virginica
148  5.4  Iris-virginica
149  5.1  Iris-virginica

[150 rows x 2 columns]

In [47]:
#calculating parent entropy

In [48]:
pe=round(-50/150*math.log(50/150,2)-50/150*math.log(50/150,2)-50/150*math.log(50/150,2),2)

In [49]:
pe

1.58

In [50]:
df.sort_values(by='PL',ascending=True,inplace=True)

C:\Users\tuhin\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [51]:
df #sorted dataframe

PL         Species
22   1.0     Iris-setosa
13   1.1     Iris-setosa
14   1.2     Iris-setosa
35   1.2     Iris-setosa
36   1.3     Iris-setosa
..   ...             ...
131  6.4  Iris-virginica
105  6.6  Iris-virginica
117  6.7  Iris-virginica
122  6.7  Iris-virginica
118  6.9  Iris-virginica

[150 rows x 2 columns]

### Spliting the data on the basis of PL column

In [141]:
#creating a function for calculating entropy then weighted entropy then information gain for each value

In [102]:
def findd(n):
    df1=df[df['PL']>n]
    df2=df[df['PL']<=n]
    
    if((df1[df1['Species']=='Iris-versicolor']['Species'].shape[0]>0) & (df1[df1['Species']=='Iris-virginica']['Species'].shape[0]>0) & (df1[df1['Species']=='Iris-setosa']['Species'].shape[0]>0) & (df2[df2['Species']=='Iris-versicolor']['Species'].shape[0]==0) & (df2[df2['Species']=='Iris-virginica']['Species'].shape[0]==0) & (df2[df2['Species']=='Iris-setosa']['Species'].shape[0]>0)):
        
        E1=round(-(df1[df1['Species']=='Iris-virginica'].shape[0])/(df1.shape[0])*math.log((df1[df1['Species']=='Iris-virginica'].shape[0])/(df1.shape[0]),2)-(df1[df1['Species']=='Iris-versicolor'].shape[0])/(df1.shape[0])*math.log((df1[df1['Species']=='Iris-versicolor'].shape[0])/(df1.shape[0]),2)-(df1[df1['Species']=='Iris-setosa'].shape[0])/(df1.shape[0])*math.log((df1[df1['Species']=='Iris-setosa'].shape[0])/(df1.shape[0]),2),2)
        E2=round(-(df2[df2['Species']=='Iris-setosa'].shape[0])/(df2.shape[0])*math.log((df2[df2['Species']=='Iris-setosa'].shape[0])/(df2.shape[0]),2),2)
        
    elif((df1[df1['Species']=='Iris-versicolor']['Species'].shape[0]>0) & (df1[df1['Species']=='Iris-virginica']['Species'].shape[0]>0) & (df1[df1['Species']=='Iris-setosa']['Species'].shape[0]==0) & (df2[df2['Species']=='Iris-versicolor']['Species'].shape[0]==0) & (df2[df2['Species']=='Iris-virginica']['Species'].shape[0]==0) & (df2[df2['Species']=='Iris-setosa']['Species'].shape[0]>0)):
        
        E1=round(-(df1[df1['Species']=='Iris-virginica'].shape[0])/(df1.shape[0])*math.log((df1[df1['Species']=='Iris-virginica'].shape[0])/(df1.shape[0]),2)-(df1[df1['Species']=='Iris-versicolor'].shape[0])/(df1.shape[0])*math.log((df1[df1['Species']=='Iris-versicolor'].shape[0])/(df1.shape[0]),2),2)
        E2=round(-(df2[df2['Species']=='Iris-setosa'].shape[0])/(df2.shape[0])*math.log((df2[df2['Species']=='Iris-setosa'].shape[0])/(df2.shape[0]),2),2)
           
    elif((df1[df1['Species']=='Iris-versicolor']['Species'].shape[0]>0) & (df1[df1['Species']=='Iris-virginica']['Species'].shape[0]>0) & (df1[df1['Species']=='Iris-setosa']['Species'].shape[0]==0) & (df2[df2['Species']=='Iris-versicolor']['Species'].shape[0]>0) & (df2[df2['Species']=='Iris-virginica']['Species'].shape[0]==0) & (df2[df2['Species']=='Iris-setosa']['Species'].shape[0]>0)):
        
        E1=round(-(df1[df1['Species']=='Iris-versicolor'].shape[0])/(df1.shape[0])*math.log((df1[df1['Species']=='Iris-versicolor'].shape[0])/(df1.shape[0]),2)-(df1[df1['Species']=='Iris-virginica'].shape[0])/(df1.shape[0])*math.log((df1[df1['Species']=='Iris-virginica'].shape[0])/(df1.shape[0]),2),2)
        E2=round(-(df2[df2['Species']=='Iris-versicolor'].shape[0])/(df2.shape[0])*math.log((df2[df2['Species']=='Iris-versicolor'].shape[0])/(df2.shape[0]),2)-(df2[df2['Species']=='Iris-setosa'].shape[0])/(df2.shape[0])*math.log((df2[df2['Species']=='Iris-setosa'].shape[0])/(df2.shape[0]),2),2)

    elif((df1[df1['Species']=='Iris-versicolor']['Species'].shape[0]>0) & (df1[df1['Species']=='Iris-virginica']['Species'].shape[0]>0) & (df1[df1['Species']=='Iris-setosa']['Species'].shape[0]==0) & (df2[df2['Species']=='Iris-versicolor']['Species'].shape[0]>0) & (df2[df2['Species']=='Iris-virginica']['Species'].shape[0]>0) & (df2[df2['Species']=='Iris-setosa']['Species'].shape[0]>0)):
        
        E1=round(-(df1[df1['Species']=='Iris-versicolor'].shape[0])/(df1.shape[0])*math.log((df1[df1['Species']=='Iris-versicolor'].shape[0])/(df1.shape[0]),2)-(df1[df1['Species']=='Iris-virginica'].shape[0])/(df1.shape[0])*math.log((df1[df1['Species']=='Iris-virginica'].shape[0])/(df1.shape[0]),2),2)
        E2=round(-(df2[df2['Species']=='Iris-virginica'].shape[0])/(df2.shape[0])*math.log((df2[df2['Species']=='Iris-virginica'].shape[0])/(df2.shape[0]),2)-(df2[df2['Species']=='Iris-versicolor'].shape[0])/(df2.shape[0])*math.log((df2[df2['Species']=='Iris-versicolor'].shape[0])/(df2.shape[0]),2)-(df2[df2['Species']=='Iris-setosa'].shape[0])/(df2.shape[0])*math.log((df2[df2['Species']=='Iris-setosa'].shape[0])/(df2.shape[0]),2),2)     
        
    elif((df1[df1['Species']=='Iris-versicolor']['Species'].shape[0]==0) & (df1[df1['Species']=='Iris-virginica']['Species'].shape[0]>0) & (df1[df1['Species']=='Iris-setosa']['Species'].shape[0]==0) & (df2[df2['Species']=='Iris-versicolor']['Species'].shape[0]>0) & (df2[df2['Species']=='Iris-virginica']['Species'].shape[0]>0) & (df2[df2['Species']=='Iris-setosa']['Species'].shape[0]>0)):       
        E1=round(-(df1[df1['Species']=='Iris-virginica'].shape[0])/(df1.shape[0])*math.log((df1[df1['Species']=='Iris-virginica'].shape[0])/(df1.shape[0]),2),2)
        E2=round(-(df2[df2['Species']=='Iris-virginica'].shape[0])/(df2.shape[0])*math.log((df2[df2['Species']=='Iris-virginica'].shape[0])/(df2.shape[0]),2)-(df2[df2['Species']=='Iris-versicolor'].shape[0])/(df2.shape[0])*math.log((df2[df2['Species']=='Iris-versicolor'].shape[0])/(df2.shape[0]),2)-(df2[df2['Species']=='Iris-setosa'].shape[0])/(df2.shape[0])*math.log((df2[df2['Species']=='Iris-setosa'].shape[0])/(df2.shape[0]),2),2)
        
    we=(df1.shape[0]/df.shape[0])*E1+(df2.shape[0]/df.shape[0])*E2 #weighted entropy
    
    ig=round(pe-we,2) #infromation gain
    
    return ig
     

In [145]:
findd(1.1) #testing the function

0.02

In [104]:
information_gain=[]

for n in df['PL']:  
    try:
    
        ig=findd(n)  
        information_gain.append(ig)
        
    except:
        
        information_gain.append(np.nan)

In [106]:
df['IG']=information_gain #creating a new column

C:\Users\tuhin\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [ ]:
#handing null values

In [134]:
df[df['IG'].isnull()] #checking for null values

PL         Species  IG
118  6.9  Iris-virginica NaN

In [116]:
df1=df[df['PL']>=6.9]
df2=df[df['PL']<6.9]

In [118]:
df1

PL         Species  IG
118  6.9  Iris-virginica NaN

In [119]:
df2

PL         Species    IG
22   1.0     Iris-setosa  0.01
13   1.1     Iris-setosa  0.02
14   1.2     Iris-setosa  0.04
35   1.2     Iris-setosa  0.04
36   1.3     Iris-setosa  0.12
..   ...             ...   ...
107  6.3  Iris-virginica  0.05
131  6.4  Iris-virginica  0.04
105  6.6  Iris-virginica  0.03
117  6.7  Iris-virginica  0.01
122  6.7  Iris-virginica  0.01

[149 rows x 3 columns]

In [123]:
E1=round(-1/1*math.log(1/1,2),2)
E1

-0.0

In [120]:
E2=round(-50/149*math.log(50/149,2)-50/149*math.log(50/149,2)-49/149*math.log(49/149,2),2)
E2

1.58

In [124]:
#weighted entropy of n>=6.9

we=round(1/150*E1+149/150*E2,2)
we

1.57

In [126]:
i_g=round(pe-we,2)
i_g

0.01

In [136]:
df['IG'].fillna(i_g,inplace=True)

C:\Users\tuhin\Anaconda3\lib\site-packages\pandas\core\generic.py:6287: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._update_inplace(new_data)


In [ ]:
#final dataset with information gain column

In [137]:
df

PL         Species    IG
22   1.0     Iris-setosa  0.01
13   1.1     Iris-setosa  0.02
14   1.2     Iris-setosa  0.04
35   1.2     Iris-setosa  0.04
36   1.3     Iris-setosa  0.12
..   ...             ...   ...
131  6.4  Iris-virginica  0.04
105  6.6  Iris-virginica  0.03
117  6.7  Iris-virginica  0.01
122  6.7  Iris-virginica  0.01
118  6.9  Iris-virginica  0.01

[150 rows x 3 columns]

In [138]:
df.isnull().sum() #no nul, value detected 

PL         0
Species    0
IG         0
dtype: int64

In [139]:
df['IG'].max()

0.91

In [140]:
df[df['IG']==0.91]

PL      Species    IG
24  1.9  Iris-setosa  0.91
44  1.9  Iris-setosa  0.91

## Hence the first spliting criteria should be PL>1.9